crates/apollo-parser/src/lexer/mod.rs

```diff
@@ -54,16 +54,75 @@ impl Lexer {
     }
 
     /// Get a reference to the lexer's tokens.
-    pub(crate) fn tokens(&self) -> &[Token] {
+    pub fn tokens(&self) -> &[Token] {
         self.tokens.as_slice()
     }
 
     /// Get a reference to the lexer's tokens.
-    pub(crate) fn errors(&self) -> &[Error] {
+    pub fn errors(&self) -> &[Error] {
         self.errors.as_slice()
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct LexerIterator<'a> {
```

Comment:
Since we are taking the streaming route, I think we should go all the way and have the Lexer take a `Read` and output a `Write`, the way a through stream would.

Reply:
why a `Write`? The output is not meant to be written to a byte buffer, it produces tokens to be consumed by the parser.
For `Read`, it is a bit tricky, because the next step in optimizing things would be to produce tokens that reference input data instead of copying it, so we quickly get into issues like trying to refill the buffer while we still hold some tokens.
Can we assume that the input data will always fit into memory? Are you planning for other input types (like ropes structures, that would make sense for editors)?